src/core/lib/channel/connected_channel.cc

```diff
@@ -203,39 +242,606 @@ static void connected_channel_get_channel_info(
+          message_to_send_ = std::move(**p);
+          send_message_state_ = SendMessageToTransport{};
+          send_message_.send_message = true;
+          batch_payload_.send_message.send_message =
+              message_to_send_->payload();
+          batch_payload_.send_message.flags = message_to_send_->flags();
+        } else {
+          GPR_ASSERT(!absl::holds_alternative<Closed>(send_message_state_));
+          client_trailing_metadata_ =
+              GetContext<FragmentAllocator>()->MakeClientMetadata();
+          send_message_state_ = Closed{};
+          send_message_.send_trailing_metadata = true;
+          batch_payload_.send_trailing_metadata.send_trailing_metadata =
+              client_trailing_metadata_.get();
+          // DO NOT SUBMIT: figure this field out
+          batch_payload_.send_trailing_metadata.sent = nullptr;
+        }
+        IncrementRefCount("send_message");
+        send_message_waker_ = Activity::current()->MakeOwningWaker();
+        push_send_message_ = true;
+        SchedulePush();
+      }
+    }
+    if (auto* pending =
+            absl::get_if<PendingReceiveMessage>(&recv_message_state_)) {
+      if (pending->received) {
+        if (pending->payload.has_value()) {
+          if (grpc_call_trace.enabled()) {
+            gpr_log(GPR_INFO,
+                    "%sRecvMessageBatchDone: received payload of %" PRIdPTR
+                    " bytes",
+                    recv_message_waker_.ActivityDebugTag().c_str(),
+                    pending->payload->Length());
+          }
+          recv_message_state_ = server_to_client_messages_->Push(
+              GetContext<FragmentAllocator>()->MakeMessage(
+                  std::move(*pending->payload), pending->flags, nullptr));
+        } else {
+          if (grpc_call_trace.enabled()) {
+            gpr_log(GPR_INFO, "%sRecvMessageBatchDone: received no payload",
+                    recv_message_waker_.ActivityDebugTag().c_str());
+          }
+          recv_message_state_ = Closed{};
+          absl::exchange(server_to_client_messages_, nullptr)->Close();
+        }
+      }
+    }
+    if (absl::exchange(queued_initial_metadata_, false)) {
+      server_initial_metadata_latch_->Set(server_initial_metadata_.get());
+    }
+    if (!absl::holds_alternative<PipeSender<MessageHandle>::PushType>(
+            recv_message_state_) &&
+        absl::exchange(queued_trailing_metadata_, false)) {
+      if (grpc_call_trace.enabled()) {
+        gpr_log(GPR_INFO,
+                "%sPollConnectedChannel: finished request, returning: {%s}; "
+                "active_ops: %s",
+                Activity::current()->DebugTag().c_str(),
+                server_trailing_metadata_->DebugString().c_str(),
+                ActiveOpsString().c_str());
+      }
+      finished_ = true;
+      return ServerMetadataHandle(std::move(server_trailing_metadata_));
+    }
+    if (absl::holds_alternative<Idle>(recv_message_state_)) {
```

Comment:
As per our discussion, we probably want to add a promise to the call context to allow the transport to know when the surface has requested a read.

We may also want to use this same mechanism for send_message instead of the callback that it has now.